MAILBOX-359 EventFactory should expose builders for events

chibenwa opened the pull request:
Following https://github.com/linagora/james-project/pull/2045#discussion_r242102511 here are builders in EventFactory for Mailbox Events.

This is part of a larger refactoring whose step 2 gonna feature **Dispatcher** removal: Event emitters will directly emit their events using these builders on the DelegatingListener, which will later be the EventBus...

mbaechler commented on mailbox/store/src/main/java/org/apache/james/mailbox/store/event/EventFactory.java:
```diff
@@ -41,7 +41,56 @@
+import com.google.common.base.Preconditions;
+
 public class EventFactory {
+    public static class MailboxAddedBuilder {
```
what about staged builder ?

chibenwa replied:
Staged builder are not good at proposing alrenative build options IMO...

mbaechler replied:
almost all your parameter are mandatory, it makes sense to use staged builder in this context.

mbaechler commented on mailbox/store/src/main/java/org/apache/james/mailbox/store/event/MailboxEventDispatcher.java:
```diff
@@ -56,23 +56,20 @@
     }
 
-    private MailboxEventDispatcher(MailboxListener listener, EventFactory eventFactory) {
+    private MailboxEventDispatcher(MailboxListener listener) {
```
you probably don't need this constructor anymore

chibenwa replied:
We do still need regisiter "MailboxListener" in tests.

IMO we can add the "delegating" layer in the tests

mbaechler commented on mailbox/api/src/main/java/org/apache/james/mailbox/MessageMoveEvent.java:
```diff
@@ -69,11 +70,16 @@ public Builder messageId(Collection<MessageId> messageIds) {
         }
 
-        public MessageMoveEvent build() {
+        public Optional<MessageMoveEvent> build() {
```
I don't like the builder taking the decision about building the event or not. I prefer the `isNoop` thing

chibenwa replied:
I will remove this commit.

chibenwa commented:
Well IMO compusition here will mean:
 - Introducing an intermediate POJO that you could later on use to build your final event like:

```
class MailboxInfo {
   private final User user;
   private final SessionId sessionId;
   private final MailboxId mailboxId;
   private final MailboxPath mailboxPath;
}
```

Not to mention that this complexity is pushed to the caller who will need to do:

```
EventFactory.added()
   .mailboxInfo(MailboxInfo.builder()
        .mailbox(mailbox)
        .session(session)
        .build())
    .metaData(MetaDataInfo.builder()
        .metaData(metaData)
        .build())
    .build();
```

and not

```
EventFactory.added()
        .mailbox(mailbox)
        .session(session)
        .metaData(metaData)
        .build());
```

Which is simpler...

blackheaven commented on mailbox/store/src/main/java/org/apache/james/mailbox/store/event/EventFactory.java:
```diff
@@ -34,74 +29,306 @@
+
+        public MailboxListener.MailboxDeletion build() {
+            mailboxEventChecks();
+            Preconditions.checkState(quotaRoot != null, "Field `quotaRoot` is compulsory");
```
`Preconditions.checkNotNull`?

chibenwa replied:
Well IMO that's a state stuff: you did not set the state of your builder well which result in something being null.

If you present it like this, the state information becomes more important IMO

chibenwa commented:
Was :green_apple: 

test this please

openpaas-james commented:
Build triggered. sha1 is merged.

chibenwa commented:
test this please

chibenwa commented:
Merged
